InterpolatedStringHandlerArgumentAttribute Decoding

333fred opened the pull request:
Validate and decode InterpolatedStringHandlerArgumentAttribute from source and from metadata. This doesn't implement call-site validation, just errors on attribute usage and metadata decoding.

333fred commented:
@chsienki @AlekseyTs for review.

chsienki commented on src/Compilers/Core/Portable/MetadataReader/PEModule.cs:
```diff
@@ -1841,7 +1870,7 @@ private static bool CrackStringAndIntInAttributeValue(out StringAndInt value, re
         }
 
-        internal static bool CrackStringInAttributeValue(out string value, ref BlobReader sig)
+        internal static bool CrackStringInAttributeValue(out string? value, ref BlobReader sig)
```
I would assume the return value accurately reflects the null status of the string, or is it possible to return true and be null? If its accurate consider using `[NotNullWhen(true)]` as above

chsienki commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs:
```diff
@@ -702,6 +704,82 @@ private static FlowAnalysisAnnotations DecodeFlowAnalysisAttributes(PEModule mod
+                        break;
+
+                    default:
+                        var param = parameters.FirstOrDefault(static (p, name) => string.Equals(p.Name, name, StringComparison.Ordinal), name);
```
Given that this is the only place we're using FirstOrDefault, is it worth just putting the implementation here, avoiding the lambda, and probably isnt any more lines of code as you don't need the below checks?

333fred replied:
I don't think so. I prefer it the way I wrote it.

chsienki commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:
```diff
@@ -1048,6 +1057,199 @@ bool needsReporting()
+            ArrayBuilder<ParameterSymbol?> parameters;
+            if (constructorArgument.Kind == TypedConstantKind.Primitive)
+            {
+                if (decodeName(constructorArgument, ref arguments) is (int ordinal, var parameter))
```
Consider using `is not` and inverting the conditional. 


```csharp
if (decodeName(constructorArgument, ref arguments) is not (int ordinal, var parameter))
{
    return;
}
parameterOrdinals = ImmutableArray.Create(ordinal);
parameters = ArrayBuilder<ParameterSymbol?>.GetInstance(1);
parameters.Add(parameter);
```

chsienki commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:
```diff
@@ -1048,6 +1057,199 @@ bool needsReporting()
+                    return (-1, null);
+                }
+
+                ImmutableArray<ParameterSymbol> parameters = ContainingSymbol switch
```
Could do this once, outside the function

chsienki commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:
```diff
@@ -1048,6 +1057,199 @@ bool needsReporting()
         }
 
+#nullable enable
+        private void DecodeInterpolatedStringHandlerArgumentAttribute(ref DecodeWellKnownAttributeArguments<AttributeSyntax, CSharpAttributeData, AttributeLocation> arguments, Binder binder, BindingDiagnosticBag diagnostics)
```
I guess i'm suprised that we would need the binder in order to decode the attribute. I see below that we're doing overload resolution to determine if its valid or not. Would an alternative be to just decode the raw strings here? Then wherever we actually use them in the binder, do the actual resolution etc? We can always get back to them from the containing symbol I think? (they always apply to other parameters in the same list as the one the attribute is on?)

333fred replied:
> Would an alternative be to just decode the raw strings here?

We could, yes, but then you don't get validation of the attribute at the attribute definition location. This seems suboptimal to me.

chsienki replied:
Hmm, yeah thats a bummer. Just seems like we're doing a lot of work here that as a caller I woulnd't expect to happen. Do we have anywhere else where the attribute itself can error based on semantics of other code?

---
In reply to: [647835991](https://github.com/dotnet/roslyn/pull/53932#discussion_r647835991) [](ancestors = 647835991)

chsienki replied:
Hmm. Could we not just store the location along with the cracked data? So we can report on it if it fails later on?

333fred replied:
> So we can report on it if it fails later on?

This doesn't help if you never use it in this project (as I would expect many to do, since these will mostly be used in libraries). Unless Aleksey has a different opinion or suggestion for how to approach this, I believe this is the right way to go.

333fred replied:
After discussion with Aleksey, we've decided that there's too much potential for issue with missing references, retargeting, and langversion changes to move forward with this for now. I'll remove the binder code.

chsienki commented:
Done with review pass (iteration 4)

333fred commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:
```diff
@@ -300,21 +300,21 @@ private static IEnumerable<string> ReadLanguagesFromAttribute(ref BlobReader arg
                         {
-                            return SpecializedCollections.SingletonEnumerable(firstLanguageName);
+                            return firstLanguageName == null ? SpecializedCollections.EmptyEnumerable<string>() : SpecializedCollections.SingletonEnumerable(firstLanguageName);
                         }
```
Users already had to deal with the empty list as a possible return. I don't think this really introduces anything new.

AlekseyTs commented:
>         private static IEnumerable<string> ReadLanguagesFromAttribute(ref BlobReader argsReader)

How changes in this fucntion are related to the goal of the PR?

---
Refers to: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:296 in b8ea696. [](commit_id = b8ea69638a74c4cfaf061454710e9dd8e22e89f8, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/MetadataReader/PEModule.cs:
```diff
@@ -1186,6 +1189,31 @@ internal bool HasIsByRefLikeAttribute(EntityHandle token)
+            }
+            else if (TryExtractStringArrayValueFromAttribute(targetAttribute.Handle, out var paramNames))
+            {
+                Debug.Assert(targetAttribute.SignatureIndex == 1);
```
>Debug.Assert(targetAttribute.SignatureIndex == 1); [](http://example.com/codeflow?start=16&length=50)

It feels like the Assert is somewhat misplaced. Should we move it before we ```TryExtractStringArrayValueFromAttribute```?

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -6656,4 +6656,32 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
+    <value>Parameter to interpolated string handler conversion occurs after handler parameter</value>
+  </data>
+  <data name="ERR_CannotUseSelfAsInterpolatedStringHandlerArgument" xml:space="preserve">
+    <value>Interpolated string handler arguments cannot refer to themselves.</value>
```
Is this error about referring to the parameter in the attribute applied to the same parameter? Perhaps it would be clearer to say something like: "cannot be used to create a handler for this parameter"?

333fred replied:
I tried to make it a bit more clear. Let me know if you think it still needs work.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -2020,6 +2020,13 @@
 
   <Node Name="BoundInterpolatedStringHandlerPlaceholder" Base="BoundValuePlaceholderBase"/>
 
+  <!-- Type expression placeholder for the arguments to the constructor call for an interpolated string handler
```
>Type expression [](http://example.com/codeflow?start=7&length=15)

"Type expression" sounds a lot like ```BoundTypeExpression```, but I believe this node doesn't represent a type reference. Because of that the comment feels confusing to me. I suggest saying "An expression" or "A typed expression" instead.

333fred commented on src/Compilers/Core/Portable/MetadataReader/PEModule.cs:
```diff
@@ -1186,6 +1189,31 @@ internal bool HasIsByRefLikeAttribute(EntityHandle token)
+            else if (TryExtractStringArrayValueFromAttribute(targetAttribute.Handle, out var paramNames))
+            {
+                Debug.Assert(targetAttribute.SignatureIndex == 1);
+                return paramNames.NullToEmpty();
```
>  Basically I think we should be able to judge from the result of GetInterpolatedStringHandlerArgumentAttributeValues if the attribute was missing or malformed.

I originally had it as `default` if missing/malformed, but thinking about it more I realized I didn't have a reason to need to understand the difference. Unless we want to actually error at the use site when we encounter a malformed attribute from metadata (which I didn't see the need to do), it feels like having the distinction really doesn't buy us much of anything.

333fred replied:
I'll record whether the data has errors in a separate bit in the packed flags.

333fred commented:
>         private static IEnumerable<string> ReadLanguagesFromAttribute(ref BlobReader argsReader)

They resulted from me nullable enabling the decoding functions (which I did because I had missed a case in the attribute decoding and was suprised I didn't get a warning about it).

---
In reply to: [857917744](https://github.com/dotnet/roslyn/pull/53932#issuecomment-857917744) [](http://example.com/codeflow?ancestors=857917744,857917744)

---
Refers to: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:296 in b8ea696. [](commit_id = b8ea69638a74c4cfaf061454710e9dd8e22e89f8, deletion_comment = False)

AlekseyTs commented:
>         private static IEnumerable<string> ReadLanguagesFromAttribute(ref BlobReader argsReader)

> They resulted from me nullable enabling the decoding functions (which I did because I had missed a case in the attribute decoding and was suprised I didn't get a warning about it).

Since the change goes beyond just adding/adjusting annotations. I think the change should be taken out of this PR and out of this feature branch. If we need to suppress nulable warnings in this function to do this, that is what we should do, I think.

---
In reply to: [857971951](https://github.com/dotnet/roslyn/pull/53932#issuecomment-857971951) [](http://example.com/codeflow?ancestors=857971951,857917744)

---
Refers to: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:296 in b8ea696. [](commit_id = b8ea69638a74c4cfaf061454710e9dd8e22e89f8, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs:
```diff
@@ -702,6 +704,82 @@ private static FlowAnalysisAnnotations DecodeFlowAnalysisAttributes(PEModule mod
+                MethodSymbol { Parameters: var p } => p,
+                PropertySymbol { Parameters: var p } => p,
+                _ => throw ExceptionUtilities.UnexpectedValue(ContainingSymbol)
+            };
```
It looks like we already have ```internal static ImmutableArray<ParameterSymbol> GetParameters(this Symbol member)``` helper that can be used here.

AlekseyTs replied:
BTW, are we going to crash on an event accesser parameter here?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs:
```diff
@@ -702,6 +704,82 @@ private static FlowAnalysisAnnotations DecodeFlowAnalysisAttributes(PEModule mod
+                switch (name)
+                {
+                    case null:
+                    case { Length: 0 } when ContainingSymbol.IsStatic:
```
>when ContainingSymbol.IsStatic: [](http://example.com/codeflow?start=39&length=31)

I guess I am missing what static-ness of a container has to do with validity of the data? Is there a relevant spec?

AlekseyTs replied:
Perhaps the Length check confused me as I thought we are dealing with an array. Consider comparing with an epty string instead, for clarity. Below as well.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceClonedParameterSymbol.cs:
```diff
@@ -164,6 +164,8 @@ internal override ImmutableHashSet<string> NotNullIfParameterNotNull
             get { return ImmutableHashSet<string>.Empty; }
         }
 
+        internal override ImmutableArray<int> InterpolatedStringHandlerArgumentIndexes => _originalParam.InterpolatedStringHandlerArgumentIndexes;
```
>_originalParam.InterpolatedStringHandlerArgumentIndexes; [](http://example.com/codeflow?start=90&length=56)

Why is this needed? And why do we assume inxes won't change?

333fred replied:
Cloned parameter symbols share attribute data, so they shouldn't change.

AlekseyTs replied:
> Cloned parameter symbols share attribute data, so they shouldn't change.

Perhaps attribute applications shouldn't change, but the meaning  of attributes can. It is quite possible that for other well-known attributes it isn't an issue because they are self-contained. However for this attribute, surrounding parameters and parameter ordinals are significant, etc. So, I would say we should reinterpret the attribute application rather than take it as is. But the question is whether there is a scenario where this particular symbol is required to expose this information at all. If not, then why return anything but a value indicating - fail creating a handler at the call site. Anything other than an "error" value would require thorough testing for this code path.

333fred replied:
I'll make these throw for now and mark with a prototype, and we'll see if the indexes are needed during lowering in the next PR. I have a hunch they might be.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Wrapped/WrappedParameterSymbol.cs:
```diff
@@ -169,6 +169,8 @@ public override string GetDocumentationCommentXml(CultureInfo preferredCulture =
             return _underlyingParameter.GetDocumentationCommentXml(preferredCulture, expandIncludes, cancellationToken);
         }
 
+        internal override ImmutableArray<int> InterpolatedStringHandlerArgumentIndexes => _underlyingParameter.InterpolatedStringHandlerArgumentIndexes;
```
>internal override ImmutableArray<int> InterpolatedStringHandlerArgumentIndexes => _underlyingParameter.InterpolatedStringHandlerArgumentIndexes; [](http://example.com/codeflow?start=8&length=144)

I would prefer overriding this property only in leaf classes. Otherwise, it is hard to predict whether this implementation is appropriate for all derived types now and in the future.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:
```diff
@@ -746,6 +751,10 @@ internal override void DecodeWellKnownAttribute(ref DecodeWellKnownAttributeArgu
                 arguments.GetOrCreateData<ParameterWellKnownAttributeData>().HasEnumeratorCancellationAttribute = true;
                 ValidateCancellationTokenAttribute(arguments.AttributeSyntaxOpt, (BindingDiagnosticBag)arguments.Diagnostics);
             }
+            else if (attribute.IsTargetAttribute(this, AttributeDescription.InterpolatedStringHandlerArgumentAttribute))
```
>IsTargetAttribute [](http://example.com/codeflow?start=31&length=17)

Since we need to deal with two different signatures here, I think we should use GetTargetAttributeSignatureIndex directly rather than trying to infer what signature we are dealing with based on values.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:
```diff
@@ -1048,6 +1057,199 @@ bool needsReporting()
+                    PropertySymbol { Parameters: var p } => p,
+                    MethodSymbol { Parameters: var p } => p,
+                    _ => throw ExceptionUtilities.UnexpectedValue(ContainingType)
+                };
```
Same comments as for PE implementation.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:
```diff
@@ -1048,6 +1057,199 @@ bool needsReporting()
+                {
+                    if (parameter.Name.Equals(name, StringComparison.Ordinal))
+                    {
+                        if ((object)parameter == this)
```
>if ((object)parameter == this) [](http://example.com/codeflow?start=24&length=30)

Are we checking this error condition for PE?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:
```diff
@@ -1048,6 +1057,199 @@ bool needsReporting()
+                    _ => throw ExceptionUtilities.UnexpectedValue(ContainingType)
+                };
+
+                foreach (var parameter in parameters)
```
>foreach (var parameter in parameters) [](http://example.com/codeflow?start=16&length=37)

It looks like for similar task for PEParameterSymbol we are using FirstOrDefault helper

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:
```diff
@@ -1048,6 +1057,199 @@ bool needsReporting()
+            foreach (var parameter in parameters)
+            {
+                var argumentType = parameter?.Type ?? ContainingType;
+                analyzedArguments.Arguments.Add(new BoundInterpolatedStringArgumentPlaceholder(arguments.AttributeSyntaxOpt!, argumentType, hasErrors: argumentType.IsErrorType()));
```
>BoundInterpolatedStringArgumentPlaceholder [](http://example.com/codeflow?start=52&length=42)

I am curious how ```BoundInterpolatedStringArgumentPlaceholder``` behaves as a target for a ref/out parameter. It is not a variable by default. Are we testing this?

333fred replied:
We have tests for this side in this PR, but we're also ignoring any diagnostics that come up from this. Regardless, since I'm removing this code it'll be tested in more depth in the next PR.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:
```diff
@@ -1048,6 +1057,199 @@ bool needsReporting()
+            else
+            {
+                // Could not find a constructor on '{0}' that matches the parameters specified in the InterpolatedStringHandlerArgumentAttribute.
+                diagnostics.Add(ErrorCode.ERR_CouldNotFindApplicableInterpolatedStringHandlerConstructor, errorLocation, handlerType);
```
>diagnostics.Add(ErrorCode.ERR_CouldNotFindApplicableInterpolatedStringHandlerConstructor, errorLocation, handlerType); [](http://example.com/codeflow?start=16&length=118)

I think we should worry whether there are things that are specific to this particular location that could cause a failure that might not occur at a different location (missing references, language version, etc.)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:
```diff
@@ -1048,6 +1057,199 @@ bool needsReporting()
+                errorName: handlerType.Name,
+                errorLocation,
+                suppressResultDiagnostics: true,
+                BindingDiagnosticBag.Discarded,
```
>BindingDiagnosticBag.Discarded [](http://example.com/codeflow?start=16&length=30)

When I see us binding with ```BindingDiagnosticBag.Discarded``` and then reporting a diagnostics based on the result, that is a red flag for me. Used asembly references are not tracked and it is quite possible that things break after they are removed because compiler says they are not used.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:
```diff
@@ -5227,5 +5227,950 @@ .locals init (CustomHandler1 V_0)
+            var vbCode = @"
+Imports System.Runtime.CompilerServices
+Public Class C
+    Public Sub M(i As Integer, <InterpolatedStringHandlerArgument({ Nothing })> c As CustomHandler)
```
>{ Nothing } [](http://example.com/codeflow?start=66&length=11)

Did you mean to  use a single element array here? Or ```CStr()Nothing)```?

333fred replied:
Yes, this is intentionally an array with 1 element, that element being `Nothing`.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:
```diff
@@ -5227,5 +5227,950 @@ .locals init (CustomHandler1 V_0)
+        }
+
+        [Fact]
+        public void InterpolatedStringHandlerArgumentAttributeError_ReferenceSelf()
```
>InterpolatedStringHandlerArgumentAttributeError_ReferenceSelf [](http://example.com/codeflow?start=20&length=61)

Are we testing this scenario from metadata?

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:
```diff
@@ -5227,5 +5227,950 @@ .locals init (CustomHandler1 V_0)
+            var vbCode = @"
+Imports System.Runtime.CompilerServices
+Public Class C
+    Public Shared Sub M(<InterpolatedStringHandlerArgument({ """" })> c As CustomHandler)
```
>{ """" } [](http://example.com/codeflow?start=59&length=8)

Are we testing single value overload?

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:
```diff
@@ -5227,5 +5227,950 @@ .locals init (CustomHandler1 V_0)
+        }
+
+        [Fact]
+        public void InterpolatedStringHandlerArgumentAttributeWarn_ParameterAfterHandler()
```
>InterpolatedStringHandlerArgumentAttributeWarn_ParameterAfterHandler [](http://example.com/codeflow?start=20&length=68)

Consider adding a metadata test too.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:
```diff
@@ -5227,5 +5227,950 @@ .locals init (CustomHandler1 V_0)
+
+public partial struct CustomHandler
+{
+    private CustomHandler(int literalLength, int formattedCount, int i) : this() {}
```
>private CustomHandler(int literalLength, int formattedCount, int i) : this() {} [](http://example.com/codeflow?start=4&length=79)

Accessibility is another aspect that is location specific. For example, internal/protected constrcutor might be accessible from where M will be called.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:
```diff
@@ -5227,5 +5227,950 @@ .locals init (CustomHandler1 V_0)
+            var handler = GetCustomHandlerType("CustomHandler", "partial struct", useBoolReturns: true);
+
+            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerArgumentAttribute, handler }, parseOptions: TestOptions.RegularPreview);
+            CompileAndVerify(comp, sourceSymbolValidator: verifier, symbolValidator: verifier);
```
>CompileAndVerify [](http://example.com/codeflow?start=12&length=16)

Consider VerifyDiagnostics on result of CompileAndVerify calls. This helps catch unexpected warnings, etc.

AlekseyTs commented:
Done with review pass (commit 5).

333fred commented:
>         private static IEnumerable<string> ReadLanguagesFromAttribute(ref BlobReader argsReader)

Filed [https://github.com/dotnet/roslyn/issues/53994](https://github.com/dotnet/roslyn/issues/53994) to track.

---
In reply to: [857980418](https://github.com/dotnet/roslyn/pull/53932#issuecomment-857980418) [](http://example.com/codeflow?ancestors=857980418,857917744)

---
Refers to: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:296 in b8ea696. [](commit_id = b8ea69638a74c4cfaf061454710e9dd8e22e89f8, deletion_comment = False)

333fred commented:
@AlekseyTs @chsienki addressed your feedback.

---
In reply to: [858138258](https://github.com/dotnet/roslyn/pull/53932#issuecomment-858138258) [](http://example.com/codeflow?ancestors=858138258,858138258)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:
```diff
@@ -5227,5 +5228,1087 @@ .locals init (CustomHandler1 V_0)
+            var vbCode = @"
+Imports System.Runtime.CompilerServices
+Public Class C
+    Public Sub M(i As Integer, <InterpolatedStringHandlerArgument({ Nothing })> c As CustomHandler)
```
>{ Nothing } [](http://example.com/codeflow?start=66&length=11)

Are we testing single value overload?

333fred commented on src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs:
```diff
@@ -34,31 +35,32 @@ private enum WellKnownAttributeFlags
         {
             // Layout:
-            // |...|fffffffff|n|rr|cccccccc|vvvvvvvv|
+            // |..|fffffffff|n|rr|ccccccccc|vvvvvvvvv|
```
Rather than creating a new structure, I think the appropriate thing to do will be to move `_bits` to a long at that point. It'll take a bit of effort when merging, but I'd much rather do that than try and create a separate side-structure for just this bit.

333fred replied:
Talked with Aleksey offline, I misread his comment. I'll do this.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Attributes/WellKnownAttributeData/ParameterWellKnownAttributeData.cs:
```diff
@@ -158,5 +158,37 @@ public void AddNotNullIfParameterNotNull(string parameterName)
+            }
+        }
+
+        private bool _hasInterpolatedStringHandlerArgumentAttributeError = false;
```
>private bool _hasInterpolatedStringHandlerArgumentAttributeError = false; [](http://example.com/codeflow?start=8&length=73)

Consider if we can get rid of this separate flag similar to the suggestion for PEParameterSymbol.

AlekseyTs commented:
@333fred It looks like there are some related test failures.

AlekseyTs reviewed (approved):
LGTM (commit 9), assuming CI is passing

333fred commented:
@chsienki can you please take another look?
